### PR TITLE
Send Modal onDismiss event on iOS (Fabric) and Android

### PR DESCRIPTION
**Pull request by bernhardoj**
## Summary:

1. Modal onDismiss is not working on iOS (Fabric).
2. Modal onDismiss is currently only available on iOS. On Android, we don't have a way to know when exactly a modal is dismissed.

Currently, the onDismiss is emitted using a device event as a workaround to the RCTModalHostView unable to receive the component event as it's already unmounted when visible is false.

This PR removes the workaround and keeps RCTModalHostView mounted until the onDismiss event is emitted from the host and sends the onDismiss event on Android.

## Changelog:

[ANDROID] [ADDED] - Added support for Modal onDismiss prop
[IOS] [FIXED] - Fix onDismiss is not working on Fabric

## Test Plan:

1. Run rn-tester
2. Open the Modal example
3. The second example shows the counter for the show and dismiss count
4. Show and dismiss the modal and verify the count is incremented correctly

https://github.com/facebook/react-native/assets/50919443/108bfb26-c8f6-43b2-ac40-f0b46e48771b

### CODE REVIEW

File: packages/rn-tester/js/examples/Modal/ModalPresentation.js
Comment (cipolleschi) [inline packages/rn-tester/js/examples/Modal/ModalPresentation.js:203]: why multiselect must be set to `true`?

**Review by cipolleschi — APPROVED** [review]
It generally looks good to me. I also love that you removed more code than you added and the overall logic looks simpler to me.
Great job! 👏

**Comment by cipolleschi** [comment]
/rebase - this comment automatically rebase the PR on top of main.

File: packages/rn-tester/js/examples/Modal/ModalPresentation.js
Comment (bernhardoj) [reply packages/rn-tester/js/examples/Modal/ModalPresentation.js:203]: It's to enable RNTOption
https://github.com/facebook/react-native/blob/b507e118678d5851555f664585bb239f6dcd6022/packages/rn-tester/js/components/RNTOption.js#L40-L44

I see that `onShow` passes `multiSelect={true}` instead of `disable={false}`, so I follow it. 
https://github.com/facebook/react-native/blob/b507e118678d5851555f664585bb239f6dcd6022/packages/rn-tester/js/examples/Modal/ModalPresentation.js#L186-L190

Do you think I should change it to `disable={false}`?

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/modal/ReactModalHostManager.java
Comment (bernhardoj) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/modal/ReactModalHostManager.java:169]: I was thinking of naming it `showOrUpdateOrDismiss`, but I guess it's too long 😅.

File: packages/rn-tester/js/examples/Modal/ModalPresentation.js
Comment (cipolleschi) [reply packages/rn-tester/js/examples/Modal/ModalPresentation.js:203]: yeah, probably it is more clear and accurate. I think that `multiSelect` is the API that allows you to select multiple choices (e.g.: "Which ice cream flavour do you prefer?" you can reply with "Chocolate", "cream" and "vanilla" and leave "strawberry" out) but in our case is is true/false and it does not make sense to select both.
WDYT?

Comment (bernhardoj) [reply packages/rn-tester/js/examples/Modal/ModalPresentation.js]: > I think that multiSelect is the API that allows you to select multiple choices

This is what I thought when saw it the first time and I totally agree using `disabled` is make more sense. Updated the code.

**Comment by cipolleschi** [comment]
Quick update on this PR. The changes implemented here are making several internal jobs fail with our CI.

This is a change we do want, but it will require a little bit more work on our side to make it happen. Ho urgent is this for you?

**Comment by bernhardoj** [comment]
@cipolleschi thanks for the update. It's not an urgent one nor blocking any workflow.

File: packages/react-native/Libraries/Modal/Modal.js
Comment (sammy-SC) [inline packages/react-native/Libraries/Modal/Modal.js:176]: What is the purpose of `isRendering?

isRendering is mapped to value of `visible` prop.

Comment (bernhardoj) [reply packages/react-native/Libraries/Modal/Modal.js:176]: `isRendering` is being used to keep the Modal rendered until the `onDismiss` callback is called.

Comment (sammy-SC) [reply packages/react-native/Libraries/Modal/Modal.js:176]: What if the parent component unmounts Modal?

Comment (bernhardoj) [reply packages/react-native/Libraries/Modal/Modal.js:176]: In that case, `onDismiss` won't be called.

**Comment by bernhardoj** [comment]
@cipolleschi hi, I see that the PR has been reverted. May I know why it was reverted and is there anything I can do to help? Thanks!

**Comment by cipolleschi** [comment]
Hi there, I was about to write.
We had to revert it because we had some callsites internally that started failing over the weekend, causing an important drop in some widely used surfaces.
Sadly, the problem is that we had some Android code which was relying on the fact that the event was not fired on that platform. 🤦 

One of the problems is that the change was quite big (changing both platforms, changing public APIs,  ...). This kind of changes are always a little risky as the more code we change the higher the probability of something to go wrong.

What we can do to make it easier for these changes to stick would be to split this PR in smaller chunks and to try and not change the public APIs.
I'll try to figure out how to split it properly

**Comment by bernhardoj** [comment]
@cipolleschi hi, hope you are doing well. I would like to know if Is there any update on the split plan.

From me, I'm thinking that perhaps we can add a new props called `useDismissEventFromNativeComponent` that will use either the current ModalEventEmitter or the new approach from this PR.

The native component itself will fire `onDismiss` and `ModalEventEmitter.modalDismissed` (for iOS) events, so the conditional logic is only applied to the JS component (Modal.js).

For the split part,
1. Start adding the Android dismiss native code. We can't test the dismiss event, but we know that it works in this PR, so I think it's fine.
2. Next, we add the iOS dismiss native code. Same as above
3. Last, we apply the conditional logic in the JS file, for example
```
componentDidMount() {
  if (ModalEventEmitter) {
    this._eventSubscription = ModalEventEmitter.addListener(
      'modalDismissed',
      event => {
        if (!this.props.useDismissEventFromNativeComponent && event.modalID === this._identifier && this.props.onDismiss) {
          this.props.onDismiss();
        }
      },
    );
  }
}
```

What do you think?

**Comment by bernhardoj** [comment]
@cipolleschi hi, do you have any updates?

**Comment by cipolleschi** [comment]
Hi @bernhardoj, sorry for the late reply, but I was quite busy with other high priority stuff and I had little to no time to look into this.

For iOS, I think everything is done now:
- In the Old Architecture, `onDismiss` relies on the old EventEmitter and it works.
- In the New Architecture, `onDismiss` uses the proper approach with the events and it works.

The problem, as you might expect, comes with Android. 
Unfortunately, I don't think we can make it work properly with that platform. 
Apparently, it turns out that, internally, we have some Android code that relies on the event not being fired. Start firing it will cause major issues with the internal codebase:
- it will take a lot of work and time to merge it
- it will be reverted multiple times, basically, every time an internal app will fail due to this change.

There are 2 stepswe can pursue:
1. We create a PR in the [discussion-and-proposal repo](https://github.com/react-native-community/discussions-and-proposals) and we get the buy in from various people, including internal people, to make this happens in Android.
2. While we wait for the above, you can create a component based on `<Modal>` with the fix. Something like `<EXPModal>` and include that component as a dependency of Expensify. Even better if, rather than copying and pasting the code, you manage to have a component that extends the native classes of Modal so that when we push some improvements, your component benefits from them too.

This will unblock both you and the app while the discussion proceeds.

How does that sound like?

**Comment by bernhardoj** [comment]
> In the New Architecture, onDismiss uses the proper approach with the events and it works.

Last time I tested, the `onDismiss` doesn't work because the JS component is unmounted before receiving the event.

> In the Old Architecture, onDismiss relies on the old EventEmitter and it works.

But I think we should remove the old EventEmitter otherwise there will be double `onDismiss` call right?

> While we wait for the above, you can create a component based on <Modal> with the fix. ...

We actually use `react-native-modal`, so I think it's not possible to apply the suggestion (`react-native-modal` has `onModalHide`, but it's called too early). Do you think applying a patch while waiting for the proposal discussion is better?

@cipolleschi

**Comment by cipolleschi** [comment]
> Last time I tested, the onDismiss doesn't work because the JS component is unmounted before receiving the event.

What I meant is that, after your diff landed, and got reverted, I split the iOS part and manage to get it landed. So, the iOS side is done and it works as you (almost) did with your diff.

 > But I think we should remove the old EventEmitter otherwise there will be double onDismiss call right?
 
No, because The New Architecture does not use the old event emitter and the Old Architecture don't use the event which is Fabric only.
So there would be no double events. 

> We actually use react-native-modal, so I think it's not possible to apply the suggestion (react-native-modal has onModalHide, but it's called too early). Do you think applying a patch while waiting for the proposal discussion is better?

Yeah, that can work too! Patch-packages is a better temporary solution than forking react-native.

**Comment by bernhardoj** [comment]
Ah, I see, thanks for clarifying it. I will discuss this with the team first. Thanks for your help!

**Pull request closed without merging**
